Smt/Solver.lean

```diff
@@ -40,84 +42,197 @@ register_option smt.solver.path : String := {
-    stdin := IO.Process.Stdio.piped
-    stdout := IO.Process.Stdio.piped
-    stderr := IO.Process.Stdio.piped
-    cmd := path
-    args := match kind with
-      | Kind.cvc5    => #["-q"]
-      | Kind.z3      => #["-in"]
-      | Kind.vampire => #["--input_syntax", "smtlib2", "--output_mode", "smtcomp"]
-  }
-  let query := String.intercalate
-                 "\n" ("(exit)\n" :: "(check-sat)" :: solver.commands).reverse
-  proc.stdin.putStr query
-  -- Close stdin to signal EOF to the solver.
-  let (_, proc) ← proc.takeStdin
-  _ ← proc.wait
-  proc.stdout.readToEnd
+def checkSat : SolverT m Result := do
+  addCommand .checkSat
+  let out ← getSexp
+  let res ← match out with
+    | "sat"     => return .sat
+    | "unsat"   => return .unsat
+    | "unknown" => return .unknown
+    | _ => (throw (IO.userError s!"unexpected solver output: {repr out}") : IO Result)
+  return res
+
+/- TODO: We should probably parse the returned string into `Command`s or `String × Term`s. This is
+   bit tricky because we need to differentiate between literals and (user-defined) symbols. It
+   should be possible, however, because we store a list of all executed commands. -/
+/-- Return the model for a `sat` result. -/
+def getModel : SolverT m String := do
+  addCommand .getModel
+  -- TODO: Replace the lines below with `getSexp` once `Sexp.parse` becomes strict.
+  let state ← get
+  let mut (ops, cps) := (0, 0)
+  let mut res := ""
+  while ops = 0 ∨ ops ≠ cps do
+    let ln ← state.proc.stdout.getLine
+    ops := ops + count '(' ln
+    cps := cps + count ')' ln
+    res := res ++ ln
+  return res
+where
+  count (c : Char) (s : String) : Nat := s.data |>.filter (· = c) |>.length
+
+/-- Check if the query given so far is satisfiable and return the result. -/
+def exit : SolverT m UInt32 := do
+  addCommand .exit
+  (← get).proc.wait
 
 end Solver
+
+def sortEndsWithNat : Term → Bool
+  | .arrowT _ t    => sortEndsWithNat t
+  | .symbolT "Nat" => true
+  | _              => false
+
+def natAssertBody (t : Term) : Term :=
+  .mkApp2 (.symbolT ">=") t (.literalT "0")
+
+open Lean Term in
+/-- TODO: remove this hack once we have a tactic that replaces Nat goals with Int goals. -/
+def natConstAssert (n : String) (args : List Name) : Term → MetaM Term
+  | arrowT i@(symbolT "Nat") t => do
+    let id ← mkFreshId
+    return (forallT id.toString i
+                   (imp id.toString (← natConstAssert n (id::args) t)))
+  | arrowT a t => do
+    let id ← mkFreshId
+    return (forallT id.toString a (← natConstAssert n (id::args) t))
+  | _ => pure $ natAssertBody (applyList n args)
+  where
+    imp n t := appT (appT (symbolT "=>") (natAssertBody (symbolT n))) t
+    applyList n : List Name → Term
+      | [] => symbolT n
+      | t :: ts => appT (applyList n ts) (symbolT t.toString)
+
+open Solver Lean Term in
+def emitCommand (cmd : Command) : SolverT MetaM (Unit) := do
+  match cmd with
+  | .setLogic l                   => setLogic l
```

Comment:
Why is this block needed, as opposed to just calling `addCommand cmd`?

Reply:
Good catch! It's not needed.